flang/docs/ParallelFortranRuntime.md

```diff
@@ -0,0 +1,1987 @@
+  subroutine prif_lcobound_no_dim(coarray_handle, lcobounds)
+    type(prif_coarray_handle), intent(in) :: coarray_handle
+    integer(c_intmax_t), intent(out) :: lcobounds(:)
+  end subroutine
+end interface
+```
+**Further argument descriptions**:
+
+* **`dim`**: which codimension of the coarray to report the lower cobound of
+* **`lcobound`**: the lower cobound of the given dimension
+* **`lcobounds`**: an array of the size of the corank of the coarray handle, returns
+  the lower cobounds of the given coarray handle
+
+### `prif_ucobound`
+
+**Description**: returns the upper cobound(s) associated with a coarray handle.
+  It is the compiler's responsibility to convert to a
+  different kind if the `kind` argument to `UCOBOUND` appears.
+    
+```
+interface prif_ucobound
+  subroutine prif_ucobound_with_dim(coarray_handle, dim, ucobound)
+    type(prif_coarray_handle), intent(in) :: coarray_handle
+    integer(c_int), intent(in) :: dim
+    integer(c_intmax_t), intent(out):: ucobound
+  end subroutine
+  subroutine prif_ucobound_no_dim(coarray_handle, ucobounds)
+    type(prif_coarray_handle), intent(in) :: coarray_handle
+    integer(c_intmax_t), intent(out) :: ucobounds(:)
+  end subroutine
+end interface
+```
+**Further argument descriptions**:
+
+* **`dim`**: which codimension of the coarray to report the upper cobound of
+* **`ucobound`**: the upper cobound of the given dimension
+* **`ucobounds`**: an array of the size of the corank of the coarray handle, returns
+    the upper cobounds of the given coarray handle
+
+### `prif_coshape`
+
+**Description**: returns the sizes of codimensions of a coarray
+    
+```
+subroutine prif_coshape(coarray_handle, sizes)
+  type(prif_coarray_handle), intent(in) :: coarray_handle
+  integer(c_size_t), intent(out) :: sizes(:)
+end subroutine
+```
+**Further argument descriptions**:
+
+* **`sizes`**: an array of the size of the corank of the coarray handle, returns the
+    difference between the upper and lower cobounds + 1
+
+### `prif_image_index`
+
+**Description**: returns the index of the image, on the identified team or the
+  current team if no team is provided, identified by the cosubscripts provided
+  in the `sub` argument with the given coarray handle
+    
+```
+subroutine prif_image_index( &
+    coarray_handle, sub, team, team_number, image_index)
+  type(prif_coarray_handle), intent(in) :: coarray_handle
+  integer(c_intmax_t), intent(in) :: sub(:)
+  type(prif_team_type), intent(in), optional :: team
+  integer(c_int), intent(in), optional :: team_number
+  integer(c_int), intent(out) :: image_index
+end subroutine
+```
+**Further argument descriptions**:
+
+* **`team` and `team_number`**: optional arguments that specify a team. They
+  shall not both be present in the same call.
+* **`sub`**: A list of integers that identify a specific image in the
+  identified or current team when interpreted as cosubscripts for the provided
+  coarray handle.
+
+## Coarray Access
```

Comment:
Quoted from @jeffhammond's response:

> All communication operations are locally blocking. There is no way for the compiler to defer or aggregate synchronization.

Reply:
**LBL response:** Fortran’s put semantics don’t require remote completion until after the next image control statement, so there can be some implicit asynchrony here. That being said, _explicitly_ asynchronous communication is future work.